context/src/main/java/io/grpc/Context.java

```diff
@@ -110,20 +113,42 @@
   private static final Key<Deadline> DEADLINE_KEY = new Key<Deadline>("deadline");
 
   /**
-   * The logical root context which is {@link #current()} if no other context is bound. This context
+   * The logical root context which is the ultimate ancestor of all contexts. This context
    * is not cancellable and so will not cascade cancellation or retain listeners.
+   *
+   * <p>Never assume this is the default context for new threads, because {@link Storage} may define
+   * a default context that is different from ROOT.
    */
   public static final Context ROOT = new Context(null);
 
-  /**
-   * Currently bound context.
-   */
-  private static final ThreadLocal<Context> localContext = new ThreadLocal<Context>() {
-    @Override
-    protected Context initialValue() {
-      return ROOT;
+  private static Storage storage;
+
+  private static synchronized Storage initializeStorage() {
+    if (storage != null) {
+      return storage;
+    }
+    try {
+      Class<?> clazz = Class.forName("io.grpc.ContextStorageOverride");
+      storage = (Storage) clazz.newInstance();
+      return storage;
+    } catch (ClassNotFoundException e) {
+      log.log(Level.FINE, "Storage override doesn't exist. Using default.", e);
+    } catch (InstantiationException e) {
+      throw new RuntimeException("Failed to initialize Storage implementation", e);
+    } catch (IllegalAccessException e) {
+      throw new RuntimeException("Failed to initialize Storage implementation", e);
+    }
+    storage = new ThreadLocalContextStorage();
+    return storage;
+  }
+
+  // For testing
+  static Storage storage() {
+    if (storage == null) {
```

Comment:
ps something like this could be fun to test using @raphw's byte buddy https://github.com/raphw/byte-buddy
seems like you could set the current classloader to some child, and load a crappy context storage into that (which you created with byte buddy). Then set the classloader back in a finally (after you've tested your errors work as should)